src/count_loc.jl

```diff
@@ -69,5 +83,184 @@ end
 count_readme(table) = count_docs(table, ("readme", "readme.md"))
 
 count_julia_loc(pkg::Package, args...) = count_julia_loc(pkg.lines_of_code, args...)
+count_docstrings(pkg::Package, args...) = count_docstrings(pkg.lines_of_code, args...)
 count_docs(pkg::Package, args...) = count_docs(pkg.lines_of_code, args...)
 count_readme(pkg::Package, args...) = count_readme(pkg.lines_of_code, args...)
+
+#####
+##### Custom Julia line counting
+#####
+
+# We can't do this at the level of SyntaxNode's because we've lost whitespace & comments already
+# So we will go to GreenNode's
+
+# Avoid piracy by defining AbstractTrees definitions on a wrapper
+struct GreenNodeWrapper
+    node::JuliaSyntax.GreenNode
+    source::JuliaSyntax.SourceFile
+end
+
+function AbstractTrees.children(wrapper::GreenNodeWrapper)
+    return map(n -> GreenNodeWrapper(n, wrapper.source), JuliaSyntax.children(wrapper.node))
+end
+
+function parse_green_one(file_path)
+    file = read(file_path, String)
+    @debug(string("Parsing ", file_path))
+    parsed = JuliaSyntax.parse(JuliaSyntax.GreenNode, file; ignore_trivia=false)
+    return GreenNodeWrapper(parsed[1], JuliaSyntax.SourceFile(file; filename=basename(file_path)))
+end
+
+# Module to make it easier w/r/t/ import clashes
+module CategorizeLines
+
+using JuliaSyntax: GreenNode, is_trivia, haschildren, is_error, children, span, SourceFile, source_location
+
+struct NodeSummary
+    starting_line::Int
+    ending_line::Int
+    summary::String
+end
+
+# Based on the recursive printing code for GreenNode's
+# Here, instead of printing, we collect all of the objects
+# that start on a given line.
+function categorize_lines!(d, node, source, nesting=0, pos=1)
+    starting_line, _ = source_location(source, pos)
+    v = get!(Vector{NodeSummary}, d, starting_line)
+
+    is_leaf = !haschildren(node)
+    if !is_leaf
+        new_nesting = nesting + 1
+        p = pos
+        for x in children(node)
+            categorize_lines!(d, x, source, new_nesting, p)
+            p += x.span
+        end
+        ending_line, _ = source_location(source, p)
+    else
+        ending_line = starting_line
+    end
+    push!(v, NodeSummary(starting_line, ending_line, summary(node),
+    # is_error=is_error(node), is_leaf, is_trivia=is_trivia(node), nesting
+    ))
+    return nothing
+end
+
+export NodeSummary
+end
+
+using .CategorizeLines
+
+# TODO:
+# Handle `@doc` calls?
+# What about inline comments #= comment =#?
+# Can a docstring not start at the beginning of a line?
+# Can there be multiple string nodes on the same line as a docstring?
```

Comment:
Unsure exactly what this means.  But probably. consider this:

```
julia> JuliaSyntax.parseall(JuliaSyntax.SyntaxNode, "\"doc1\" x; \"doc2 \$an_interpolation\" y")
line:col│ tree                                   │ file_name
   1:1  │[toplevel]
   1:1  │  [toplevel]
   1:1  │    [macrocall]
   1:1  │      core_@doc
   1:1  │      [string]
   1:2  │        "doc1"
   1:8  │      x
   1:10 │    [macrocall]
   1:10 │      core_@doc
   1:11 │      [string]
   1:12 │        "doc2 "
   1:18 │        an_interpolation
   1:36 │      y
```

Reply:
ah good point. Not sure how to handle this. What I do to count docstrings is I find a docstring call, take the first string node after that, and then look where that node ends (which line). Then I count all those lines as docstring lines. This handles things like
```
"""
abc
def
"""
f
```
which should count as 4 lines I think (or maybe 2?).

If someone did your example, I think that's fine to count as 1 docstring line. But something like

```
"""
abc
def
"""
function f end; """
a
b
c
"""
function g end
```
currently is handled badly:
```julia
julia> PackageAnalyzer.LineCategories("test/lines_of_code/docstrings.jl")
┌ Debug: Parsing test/lines_of_code/docstrings.jl
└ @ PackageAnalyzer ~/PackageAnalyzer.jl/src/count_loc.jl:109
1    | Docstring | """
2    | Docstring | abc
3    | Docstring | def
4    | Docstring | """
5    | Code      | f; """
6    | Code      | a
7    | Code      | b
8    | Code      | c
9    | Code      | """
10   | Code      | g
```

BTW: this case does not parse on JuliaSyntax v0.3.0, 0.3.1, or 0.3.2. Even a slightly simple variant fails:
```julia
julia> JuliaSyntax.parse(JuliaSyntax.GreenNode, """
       "
       a
       "
       f
       g
       """; ignore_trivia=false)
ERROR: ParseError:
Error: unexpected text after parsing statement
@ line 5:1
"
f
g


```

Reply:
I suspect a lot of these problems would go away if you give up on the 1:1 mapping between source files and classification, and allow each line to have multiple labels.

For the failing parse, you need `JuliaSyntax.parseall` - that's for parsing a whole file of top-level statements, not just a single statement:

```julia
julia> JuliaSyntax.parseall(JuliaSyntax.GreenNode, """
              "
              a
              "
              f
              g
              """; ignore_trivia=false)
     1:10     │[toplevel]
     1:7      │  [macrocall]
     1:0      │    core_@doc            ✔
     1:5      │    [string]
     1:1      │      "
     2:4      │      String             ✔
     5:5      │      "
     6:6      │    NewlineWs
     7:7      │    Identifier           ✔
     8:8      │  NewlineWs
     9:9      │  Identifier             ✔
    10:10     │  NewlineWs
```

Reply:
Ahh ok, I didn’t get that, thanks.

I agree it would make these decisions easier… I kinda hate to give it up though. I also don’t think that’s how any other loc tool works. Would love to see examples otherwise though.

Reply:
If you don't want to give it up that's fine.  But you could make it part of the implementation:

* Pass 1: Classify all lines into multiple categories according to which nodes touch them
* Pass 2: Reduce each line to a single category according to a precedence rule -something like `code > docstring > comment` or whatever

This would also fix the "what to do about inline comments" in an easy way.

Reply:
This is a great idea. I was also able to push it up into the traversal so I don't need to keep a list of all objects on every line, and I think having a clear rule made that easier too.

Reply:
Sounds great!